src/openforms/submissions/api/mixins.py

```diff
@@ -0,0 +1,58 @@
+from django.db import transaction
+from django.utils import timezone
+
+from rest_framework.request import Request
+from rest_framework.reverse import reverse
+
+from openforms.logging import logevent
+
+from ..models import Submission
+from ..signals import submission_complete
+from ..tasks import on_completion
+from ..tokens import submission_status_token_generator
+from ..utils import (
+    persist_user_defined_variables,
+    remove_submission_from_session,
+    remove_submission_uploads_from_session,
+)
+
+
+class SubmissionCompletionMixin:
+    request: Request
+
+    def _complete_submission(self, submission: Submission) -> str:
+        """
+        Mark the submission as completed.
+
+        This encapsulates the logic of what it means to 'complete' a submission,
+        ensuring that the relevant metadata is set and post-completation hooks trigger,
+        such as schedulign the processing via Celery.
+        """
+
+        # dispatch signal for modules to tap into
+        submission_complete.send(sender=self.__class__, request=self.request)
+
+        submission.calculate_price(save=False)
+        submission.completed_on = timezone.now()
+        submission.save()
+
+        persist_user_defined_variables(submission, self.request)
```

Comment:
Does this cause problems because the rules are evaluated twice? I think both `calculate_price` and `persist_user_defined_variables` may evaluate them and I have some memories of this causing problems :thinking:

Reply:
I don't think so, as I just moved this code out of the API endpoint into a re-usable mixin?